edb/tools/test/results.py

```diff
@@ -0,0 +1,298 @@
+
+
+def render_result(
+    file: typing.IO,
+    result: TestResult,
+) -> None:
+    _echo(file)
+
+    # cases
+    for case in result.warnings:
+        _print_case_result(file, case, 'WARNING', 'yellow')
+    for case in result.errors:
+        _print_case_result(file, case, 'ERROR', 'red')
+    for case in result.failures:
+        _print_case_result(file, case, 'FAIL', 'red')
+    for case in result.unexpected_successes:
+        _print_case_result(file, case, 'UNEXPECTED SUCCESS', 'red')
+
+    # outcome
+    if result.was_successful:
+        _echo(file, 'SUCCESS', fg='green', bold=True)
+    else:
+        _echo(file, 'FAILURE', fg='red', bold=True)
+
+    # counts
+    counts = [
+        ('tests ran', result.testsRun),
+        ('failures', len(result.failures)),
+        ('errors', len(result.errors)),
+        ('expected failures', len(result.expected_failures)),
+        ('not implemented', len(result.not_implemented)),
+        ('unexpected successes', len(result.unexpected_successes)),
+        ('skipped', len(result.skipped)),
+    ]
+    for bit, count in counts:
+        if not count:
+            continue
+        _echo(file, f'  {bit}: ', nl=False)
+        _echo(file, f'{count}', bold=True)
+
+    # running times
+    _echo(file)
+    _echo(file, f'Running times: ')
+    if result.boot_time_taken > 0.0:
+        _echo(file, '  bootstrap: ', nl=False)
+        _echo(file, _format_time(result.boot_time_taken), bold=True)
+
+    _echo(file, '  tests: ', nl=False)
+    _echo(file, _format_time(result.tests_time_taken), bold=True)
+
+    if result.boot_time_taken > 0.0:
+        _echo(file, '  total: ', nl=False)
+        _echo(
+            file,
+            _format_time(result.boot_time_taken + result.tests_time_taken),
+            bold=True,
+        )
+
+    _echo(file)
+
+
+def _result_log_dir() -> pathlib.Path:
+    dir = pathlib.Path('.') / 'build' / 'test-results'
+    dir.mkdir(exist_ok=True)
+    return dir
+
+
+def _result_log_file() -> typing.IO:
+    dir = _result_log_dir()
+    now = str(datetime.datetime.now()).replace(' ', '_')
+    return open(dir / f'{now}.json', 'w')
+
+
+def write_result(res: TestResult):
+    log_file = _result_log_file()
+
+    json.dump(dataclasses.asdict(res), log_file, indent=4)
+
+
+def read_unsuccessful() -> typing.List[str]:
+    dir = _result_log_dir()
+    results = list(dir.iterdir())
+    if not results:
+        return []
+
+    results.sort()
+    last = results[-1]
+
+    try:
+        result_dict = json.load(open(last, 'r'))
+        result = _dataclass_from_dict(TestResult, result_dict)
+    except Exception:
+        return []
+    return [
+        case.name
+        for case in result.failures
+        + result.errors
+        + result.unexpected_successes
+    ]
+
+
+def _dataclass_from_dict(klass: typing.Type, d: typing.Dict):
+    # JSON -> dataclasses is surprisingly hard.
+    # Maybe we could use a library here.
+
+    # HACK: hardcoded List[TestCase]
```

Comment:
You should be able to use our ersatz `typing_inspect` to extract the argument from the list type with `get_args`